It looks like normally, h5py keeps the *linked* file open after the *linking* file is closed on all OS's. But on Mac/Linux, the *linked* file can be deleted or truncated, but on Windows, trying to delete or truncate the *linked* file results in an error.
```
Traceback (most recent call last):
  File "C:\Users\Ryan\Documents\NWB\hdmf\tests\unit\test_io_hdf5_h5tools.py", line 996, in tearDown
    os.remove(self.path1)  # linked file may not be closed
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\Ryan\\AppData\\Local\\Temp\\tmpabnbjx67'
```

See also https://github.com/NeurodataWithoutBorders/pynwb/issues/975

This seems like a bug in h5py or just a Windows filesystem quirk, so I filed an issue ticket there: https://github.com/h5py/h5py/issues/1569

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -556,20 +556,20 @@ class HDF5IO(HDMFIO):
                 kwargs = {}
             self.__file = File(self.__path, open_flag, **kwargs)
 
-    def close(self, close_links=True):
-        """Close the file.
-
-        :param bool close_links: Whether to explicitly close all opened, linked-to files. MacOS and Linux automatically
-                                 closes opened, linked-to files when the linking file is closed, but Windows does not.
-                                 This flag allows Windows users not to close those linked-to files. Default True.
-        """
+    def close(self):
         if self.__file is not None:
             self.__file.close()
-        if close_links:
-            for obj in self.__open_links:
-                if obj:
-                    obj.file.close()
-            self.__open_links = []
+
+    def close_linked_files(self):
+        """Close all opened, linked-to files.
+
+        MacOS and Linux automatically releases the linked-to file after the linking file is closed, but Windows does
+        not, which prevents the linked-to file from being deleted or truncated.
+        """
+        for obj in self.__open_links:
+            if obj:
+                obj.file.close()
+        self.__open_links = []
 
     @docval({'name': 'builder', 'type': GroupBuilder, 'doc': 'the GroupBuilder object representing the HDF5 file'},
             {'name': 'link_data', 'type': bool,

--- a/src/hdmf/backends/hdf5/h5tools.py
+++ b/src/hdmf/backends/hdf5/h5tools.py
@@ -567,7 +567,8 @@ class HDF5IO(HDMFIO):
             self.__file.close()
         if close_links:
             for obj in self.__open_links:
-                obj.file.close()
+                if obj:
+                    obj.file.close()
             self.__open_links = []
 
     @docval({'name': 'builder', 'type': GroupBuilder, 'doc': 'the GroupBuilder object representing the HDF5 file'},

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -1025,7 +1025,7 @@ class TestCloseLinks(TestCase):
         self.assertFalse(read_foofile2.foo_link.my_data)
 
         # should be able to reopen both files
-        with HDF5IO(self.path1, mode='w', manager=_get_manager()) as new_io3:
+        with HDF5IO(self.path1, mode='a', manager=_get_manager()) as new_io3:
             new_io3.read()
 
     def test_double_close_file_with_links(self):

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -1020,14 +1020,12 @@ class TestCloseLinks(TestCase):
         with HDF5IO(self.path2, mode='a', manager=_get_manager()) as new_io1:
             read_foofile2 = new_io1.read()  # keep reference to container in memory
 
-        self.assertEqual(read_foofile2.foo_link.name, 'foo1')
+        self.assertTrue(read_foofile2.foo_link.my_data)
+        new_io1.close_linked_files()
         self.assertFalse(read_foofile2.foo_link.my_data)
 
         # should be able to reopen both files
-        with HDF5IO(self.path2, mode='a', manager=_get_manager()) as new_io2:
-            new_io2.read()
-
-        with HDF5IO(self.path1, mode='a', manager=_get_manager()) as new_io3:
+        with HDF5IO(self.path1, mode='w', manager=_get_manager()) as new_io3:
             new_io3.read()
 
     def test_double_close_file_with_links(self):
@@ -1052,9 +1050,10 @@ class TestCloseLinks(TestCase):
 
         with HDF5IO(self.path2, mode='a', manager=_get_manager()) as new_io1:
             read_foofile2 = new_io1.read()  # keep reference to container in memory
-            read_foofile2.foo_link.my_data.file.close()  # explicitly close the h5dataset
-            self.assertFalse(read_foofile2.foo_link.my_data)
-        # make sure new_io1.close() does not fail because the linked-to file is already closed
+
+        read_foofile2.foo_link.my_data.file.close()  # explicitly close the file from the h5dataset
+        self.assertFalse(read_foofile2.foo_link.my_data)
+        new_io1.close_linked_files()  # make sure this does not fail because the linked-to file is already closed
 
 
 class HDF5IOInitNoFileTest(TestCase):

--- a/tests/unit/test_io_hdf5_h5tools.py
+++ b/tests/unit/test_io_hdf5_h5tools.py
@@ -975,7 +975,7 @@ class HDF5IOMultiFileTest(TestCase):
         self.assertIsInstance(f3.get('/buckets/test_bucket2/foo_holder/foo2/my_data', getlink=True), HardLink)
 
 
-class TestOpenLinks(TestCase):
+class TestCloseLinks(TestCase):
 
     def setUp(self):
         self.path1 = get_temp_filepath()
@@ -1021,6 +1021,7 @@ class TestOpenLinks(TestCase):
             read_foofile2 = new_io1.read()  # keep reference to container in memory
 
         self.assertEqual(read_foofile2.foo_link.name, 'foo1')
+        self.assertFalse(read_foofile2.foo_link.my_data)
 
         # should be able to reopen both files
         with HDF5IO(self.path2, mode='a', manager=_get_manager()) as new_io2:
@@ -1028,6 +1029,32 @@ class TestOpenLinks(TestCase):
 
         with HDF5IO(self.path1, mode='a', manager=_get_manager()) as new_io3:
             new_io3.read()
+
+    def test_double_close_file_with_links(self):
+        # Create the first file
+        foo1 = Foo('foo1', [0, 1, 2, 3, 4], "I am foo1", 17, 3.14)
+        bucket1 = FooBucket('test_bucket1', [foo1])
+        foofile1 = FooFile(buckets=[bucket1])
+
+        # Write the first file
+        with HDF5IO(self.path1, mode='w', manager=_get_manager()) as io:
+            io.write(foofile1)
+
+        # Create the second file
+        manager = _get_manager()  # use the same manager for read and write so that links work
+        with HDF5IO(self.path1, mode='r', manager=manager) as read_io:
+            read_foofile1 = read_io.read()
+            foofile2 = FooFile(foo_link=read_foofile1.buckets[0].foos[0])  # cross-file link
+
+            # Write the second file
+            with HDF5IO(self.path2, mode='w', manager=manager) as write_io:
+                write_io.write(foofile2)
+
+        with HDF5IO(self.path2, mode='a', manager=_get_manager()) as new_io1:
+            read_foofile2 = new_io1.read()  # keep reference to container in memory
+            read_foofile2.foo_link.my_data.file.close()  # explicitly close the h5dataset
+            self.assertFalse(read_foofile2.foo_link.my_data)
+        # make sure new_io1.close() does not fail because the linked-to file is already closed
 
 
 class HDF5IOInitNoFileTest(TestCase):

